nuxeo-features/nuxeo-automation/nuxeo-automation-client/src/main/java/org/nuxeo/ecm/automation/client/adapters/AsyncSession.java

```diff
@@ -0,0 +1,289 @@
+import org.nuxeo.ecm.automation.client.Session;
+import org.nuxeo.ecm.automation.client.jaxrs.spi.Connector;
+import org.nuxeo.ecm.automation.client.jaxrs.spi.DefaultOperationRequest;
+import org.nuxeo.ecm.automation.client.jaxrs.spi.DefaultSession;
+import org.nuxeo.ecm.automation.client.jaxrs.spi.JsonMarshalling;
+import org.nuxeo.ecm.automation.client.jaxrs.spi.Request;
+import org.nuxeo.ecm.automation.client.jaxrs.util.MultipartInput;
+import org.nuxeo.ecm.automation.client.model.Blob;
+import org.nuxeo.ecm.automation.client.model.Blobs;
+import org.nuxeo.ecm.automation.client.model.OperationDocumentation;
+import org.nuxeo.ecm.automation.client.model.OperationInput;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.nuxeo.ecm.automation.client.Constants.CTYPE_REQUEST_NOCHARSET;
+import static org.nuxeo.ecm.automation.client.Constants.HEADER_NX_SCHEMAS;
+import static org.nuxeo.ecm.automation.client.Constants.REQUEST_ACCEPT_HEADER;
+
+public class AsyncSession implements Session {
+
+    /**
+     * Request providing a completable call method for convenience.
+     */
+    class CompletableRequest extends Request {
+
+        private CompletableFuture<CompletableRequest> future;
+
+        int status;
+        private Header[] headers;
+        private Object result;
+
+        public CompletableRequest(int method, String url) {
+            super(method, url, (String) null);
+        }
+
+        public CompletableRequest(int method, String url, String entity) {
+            super(method, url, entity);
+        }
+
+        public CompletableRequest(int method, String url, MultipartInput input) {
+            super(method, url, input);
+        }
+
+        @Override
+        public Object handleResult(int status, Header[] headers, InputStream stream)
+                throws RemoteException, IOException {
+            this.status = status;
+            this.headers = headers;
+            try {
+                this.result = super.handleResult(status, headers, stream);
+                future.complete(this);
+            } catch (RemoteException e) {
+                future.completeExceptionally(e);
+            }
+            return result;
+        }
+
+        AsyncSession getSession() {
+            return AsyncSession.this;
+        }
+
+        public String getHeader(String name) {
+            return Request.getHeaderValue(headers, name);
+        }
+
+        public CompletableFuture<? extends CompletableRequest> call() {
+            future = new CompletableFuture<>();
+            try {
+                getSession().getConnector().execute(this);
```

Comment:
`execute(this)` will invoke `handleResults` which will them complete the Future. Just convenience really and to avoid refactoring too much the code (which is deprecated anyway)